Tests/tests/test_ControlService.cpp

```diff
@@ -0,0 +1,798 @@
+    NiceMock<FactoriesImplementation> factoriesImplementation_;
+    Core::JSONRPC::Message            message_;
+    PluginHost::IDispatcher*          dispatcher_;
+
+    ControlServiceInitializedEventTest() :
+        ControlServiceTest()
+    {
+        EXPECT_CALL(iarmBusImplMock_, IARM_Bus_RegisterEventHandler(::testing::_, ::testing::_, ::testing::_))
+            .WillOnce(::testing::Invoke(
+                [&](const char* ownerName, IARM_EventId_t eventId, IARM_EventHandler_t handler) {
+                    EXPECT_EQ(string(ownerName), string(IARM_BUS_IRMGR_NAME));
+                    EXPECT_EQ(eventId, IARM_BUS_IRMGR_EVENT_IRKEY);
+                    controlEventHandler_ = handler;
+                    return IARM_RESULT_SUCCESS;
+                }))
+            .WillRepeatedly(::testing::Invoke(
+                [&](const char* ownerName, IARM_EventId_t eventId, IARM_EventHandler_t handler) {
+                    EXPECT_EQ(string(ownerName), string(CTRLM_MAIN_IARM_BUS_NAME));
+                    EXPECT_TRUE(isValidCtrlmRcuIarmEvent(eventId));
+                    controlEventHandler_ = handler;
+                    return IARM_RESULT_SUCCESS;
+                }));
+
+        EXPECT_EQ(string(""), plugin_->Initialize(nullptr));
+        PluginHost::IFactories::Assign(&factoriesImplementation_);
+        dispatcher_ = static_cast<PluginHost::IDispatcher*>(plugin_->QueryInterface(PluginHost::IDispatcher::ID));
+        dispatcher_->Activate(&service_);
+    }
+
+    virtual ~ControlServiceInitializedEventTest() override
+    {
+        plugin_->Deinitialize(nullptr);
+        dispatcher_->Deactivate();
+        dispatcher_->Release();
+        PluginHost::IFactories::Assign(nullptr);
+    }
+};
+
+TEST_F(ControlServiceTest, RegisteredMethods)
+{
+    EXPECT_EQ(Core::ERROR_NONE, handler_.Exists(_T("getQuirks")));
+    EXPECT_EQ(Core::ERROR_NONE, handler_.Exists(_T("getAllRemoteData")));
+    EXPECT_EQ(Core::ERROR_NONE, handler_.Exists(_T("getSingleRemoteData")));
+    EXPECT_EQ(Core::ERROR_NONE, handler_.Exists(_T("getLastKeypressSource")));
+    EXPECT_EQ(Core::ERROR_NONE, handler_.Exists(_T("getLastPairedRemoteData")));
+    EXPECT_EQ(Core::ERROR_NONE, handler_.Exists(_T("setValues")));
+    EXPECT_EQ(Core::ERROR_NONE, handler_.Exists(_T("getValues")));
+    EXPECT_EQ(Core::ERROR_NONE, handler_.Exists(_T("startPairingMode")));
+    EXPECT_EQ(Core::ERROR_NONE, handler_.Exists(_T("endPairingMode")));
+    EXPECT_EQ(Core::ERROR_NONE, handler_.Exists(_T("canFindMyRemote")));
+    EXPECT_EQ(Core::ERROR_NONE, handler_.Exists(_T("findLastUsedRemote")));
+    EXPECT_EQ(Core::ERROR_NONE, handler_.Exists(_T("checkRf4ceChipConnectivity")));
+}
+
+TEST_F(ControlServiceTest, getQuirks)
+{
+    EXPECT_EQ(Core::ERROR_NONE, handler_.Invoke(connection_, _T("getQuirks"), _T("{}"), response_));
+    EXPECT_EQ(response_, quirksResponse);
+}
+
+static ::testing::AssertionResult mockCtrlmData(const char* methodName, void* arg)
```

Comment:
Every test should be easy to read. Setup is important for understanding the test . If several tests do the same setup it can be moved elsewhere. Also if setup is large, parts of it can be moved elsewhere. But moving all setup elsewhere introduces a puzzle. In order to understand a test one needs to scroll several hundreds lines . Without having a deep knowledge on the code nobody will tell which if-else is triggered. For example, "GP502KXBG" . It's set at line 227, checked at line 479, and defined at line 55. If somebody will need to maintain this they will spend hours collecting all pieces of the puzzle. Please try to keep as much info as possible in the test.

Reply:
I moved the set up and data specific to tests within the individual tests that needed them but kept common setups in the mockCtrlmData() method

Reply:
I will have to mark this pr invalid because the tests you add don't verify anything, and their setup/expectations are in unclear places outside of the test, which makes them not maintainable.

Reply:
Refactored tests to follow the Arrange-Act-Assert flow hopefully the setup/expectations and what behaviors the tests are verifying are clearer